charms/kserve-controller/tests/unit/test_charm.py

```diff
@@ -0,0 +1,201 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+
+from unittest.mock import MagicMock
+
+import pytest
+from charmed_kubeflow_chisme.exceptions import ErrorWithStatus
+from lightkube.core.exceptions import ApiError
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops.testing import Harness
+
+from charm import KServeControllerCharm
+
+
+RBAC_PROXY_EXPECTED_LAYER = {
+    "services": {
+        "kube-rbac-proxy": {
+            "override": "replace",
+            "summary": "Kube Rbac Proxy",
+            "command": "/usr/local/bin/kube-rbac-proxy --secure-listen-address=0.0.0.0:8443 --upstream=http://127.0.0.1:8080 --logtostderr=true --v=10",
+            "startup": "enabled",
+        }
+    }
+}
+
+
+class _FakeErrorWithStatus(ErrorWithStatus):
+    def __init__(self, status_type=BlockedStatus):
+        super().__init__("err", status_type)
+
+
+class _FakeResponse:
+    """Used to fake an httpx response during testing only."""
+
+    def __init__(self, code):
+        self.code = code
+
+    def json(self):
+        return {"apiVersion": 1, "code": self.code, "message": "broken"}
+
+
+class _FakeApiError(ApiError):
+    """Used to simulate an ApiError during testing."""
+
+    def __init__(self, code=400):
+        super().__init__(response=_FakeResponse(code))
+
+
+@pytest.fixture
+def harness():
+    """Returns a harnessed charm with leader == True."""
+    harness = Harness(KServeControllerCharm)
+    harness.set_leader(True)
+    return harness
+
+
+@pytest.fixture()
+def mocked_resource_handler(mocker):
+    """Yields a mocked instance of the KubernetesResourceHAndler."""
+    mocked_resource_handler = MagicMock()
+    mocked_resource_handler_factory = mocker.patch("charm.KubernetesResourceHandler")
+    mocked_resource_handler_factory.return_value = mocked_resource_handler
+    yield mocked_resource_handler
+
+
+@pytest.fixture()
+def mocked_lightkube_client(mocker, mocked_resource_handler):
+    """Prevents lightkube clients from being created, returning a mock instead."""
+    mocked_resource_handler.lightkube_client = MagicMock()
+    yield mocked_resource_handler.lightkube_client
+
+
+@pytest.fixture()
+def mocked_gen_certs(mocker):
+    """Yields a mocked gen_certs."""
+    yield mocker.patch("charm.KServeControllerCharm.gen_certs")
+
+
```

Comment:
Doc strings in all tests from this line to the end. Linting will fail without those.